core-bundle/src/Filesystem/FilesystemItemIterator.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Filesystem;
+
+/**
+ * @implements \IteratorAggregate<int, FilesystemItem>
+ */
+class FilesystemItemIterator implements \IteratorAggregate
+{
+    /**
+     * @var iterable<FilesystemItem>
+     */
+    private iterable $listing;
+
+    /**
+     * @param iterable<FilesystemItem> $listing
+     */
+    public function __construct(iterable $listing)
+    {
+        $this->listing = $listing;
+    }
+
+    /**
+     * @param callable(FilesystemItem):bool $filter
+     */
+    public function filter(callable $filter): self
+    {
+        $listFiltered = static function (iterable $listing) use ($filter): \Generator {
+            foreach ($listing as $item) {
+                if ($filter($item)) {
+                    yield $item;
+                }
+            }
+        };
+
+        return new self($listFiltered($this->listing));
```

Comment:
Specific reason no use a bound closure?

```suggestion
        $listFiltered = function () use ($filter): \Generator {
            foreach ($this->listing as $item) {
                if ($filter($item)) {
                    yield $item;
                }
            }
        };

        return new self($listFiltered());
```

Reply:
> Specific reason no use a bound closure?

Kind of: a static function does not hold a reference to the object which would prevent garbage collection. It's probably not a huge deal here, because the object is quite small, but it's probably a good idea in general for chained iterators.

Reply:
I constructed an example where the difference is quite severe: https://3v4l.org/Ng5eU#v8.1.2 (switch to the commented version to see the difference). I had to make the object artificially large and store references to make the effect become visible, though. :slightly_smiling_face:

Reply:
I just thought that because the object only consists of the `listing` member that the memory usage would be basically the same. But I agree, if other members get added later or as a general guidline making it static shows the dependencies to the developer directly in the code. 👍